src/middleware/devtools.ts

```diff
@@ -1,167 +1,325 @@
-          connection.init(api.getState())
-        } else if (
-          message.type === 'DISPATCH' &&
-          message.payload?.type === 'IMPORT_STATE'
-        ) {
-          const actions = message.payload.nextLiftedState?.actionsById
-          const computedStates =
-            message.payload.nextLiftedState?.computedStates || []
-
-          computedStates.forEach(
-            ({ state }: { state: PartialState<S> }, index: number) => {
-              const action = actions[index] || 'No action found'
-
-              if (index === 0) {
-                connection.init(state)
-              } else {
-                savedSetState(state)
-                connection.send(action, api.getState())
-              }
-            }
-          )
+        | { type: UnknownString }
+    }
+  | { type: UnknownString }
+
+type EStateLifted =
+  { computedStates: { state: O.Partial<EStateFromDevtools> }[] }
+type EDevtoolsPrefix = string & { __isDevtoolsPrefix: true }
+type EActionToSend = { type: string } & { __isActionToSend: true }  
+type EStateFromDevtools = { __isStateFromDevtools: true }
+
+const devtoolsWindow = window as
+  & Window
+  & DevtoolsWindow
+  & { top: DevtoolsWindow }
+
+const devtoolsImpl: EDevtools =
+  (storeInitializer, _devtoolsOptions) =>
+    (parentSet, parentGet, parentStore) => {
+
+  let devtoolsOptions: EDevtoolsOptions =
+    _devtoolsOptions === undefined ? { name: undefined } :
+    typeof _devtoolsOptions === 'string' ? { name: _devtoolsOptions } :
+    _devtoolsOptions
+
+  let extensionConnector =
+    devtoolsWindow.__REDUX_DEVTOOLS_EXTENSION__ ??
+    devtoolsWindow.top.__REDUX_DEVTOOLS_EXTENSION__
+
+  if (!extensionConnector) {
+    if (process.env.NODE_ENV === 'development' && typeof window !== 'undefined') {
+      console.warn('Please install/enable Redux devtools extension')
+    }
+
+    return storeInitializer(parentSet, parentGet, parentStore)
+  }
+
+  let store: EStore & EDevtoolsStore = parentStore;
+  if (!store.devtools) store.devtools = extensionConnector.connect(devtoolsOptions);
+  store.devtools.prefix = prefixFromName(devtoolsOptions.name);
+
+  let isSilentSetState = false;
+  store.setState = (...[state, replace, action]: F.Arguments<typeof store['setState']>) => {
+    parentSet(state, replace)
+    if (!isSilentSetState) return;
+    store.devtools?.send(
+      actionToSend(action, store.devtools.prefix),
+      parentGet()
+    )
+  }
+  const silentlySetState = (state: EStateFromDevtools | O.Partial<EStateFromDevtools>) => {
+    isSilentSetState = true;
+    store.setState(state as unknown as EState);
+    isSilentSetState = false;
+  }
+
+  let initialState =
+    storeInitializer(store.setState, parentGet, store);
+
+  store.devtools.subscribe(message => {
+    switch (message.type) {
+      case 'ACTION':
+        if (!isReduxLike(store)) return;
+        if (typeof message.payload !== "string") return;
+        return parseJsonThen(message.payload, store.dispatch);
```

Comment:
This was previously done in the redux middleware...
https://github.com/pmndrs/zustand/blob/c8ae965435beb2cfb5a8bb8106e4e81712e9de25/src/middleware/redux.ts#L31-L33
So now I moved it here in it's right place.

Also just checking for `dispatch` property is unsafe because the user could have defined their own `dispatch` that is supposed to do something else, so only when the `isReduxLike` flag is set (which the `redux` middleware does) we assume `dispatch` is actual action dispatcher.

Reply:
Is this just refactoring or changing the behavior?

Reply:
Given that I realised we were sending state as the action, this might be changing the behavior, will revert.